Support python 3.11

apotterri opened the pull request:
**Please don't merge this, it needs a rebase**

Fixes #236.

dividedmind commented on _appmap/unittest.py:
```diff
@@ -12,39 +13,59 @@ def _get_test_location(cls, method_name):
+    # unittest.case._Outcome.testPartExecutor is used by all supported
+    # versions of unittest to run test cases. `isTest` will be True when
+    # the part is the actual test method, False when it's setUp or
+    # teardown.
```
It seems like you copied this comment verbatim and forgot to update it.

apotterri replied:
Oops, yeah, thanks. Fixed.

dividedmind commented on _appmap/unittest.py:
```diff
@@ -12,39 +13,59 @@ def _get_test_location(cls, method_name):
-                *args, **kwargs
-            ), testing_framework.collect_result_metadata(metadata):
+if sys.version_info[1] < 8:
+    # unittest.case._Outcome.testPartExecutor is used by all supported
```
No longer 'all supported', right?

dividedmind reviewed (changes requested):
Judging by the CI log, it seems that moving `wrapt` to `_appmap` might not have been the best idea. What is the reason to move, how was it more difficult to upgrade when it was in `appmap`? The commit message doesn't make it entirely clear.

Other than that and the comments, please add 3.11 to CI (and maybe move smoke test to that).

dividedmind commented on _appmap/importer.py:
```diff
@@ -264,12 +262,26 @@ def wrapped_find_spec(find_spec, _, args, kwargs):
+        find_spec = getattr(finder, "find_spec", None)
+        if find_spec is None:
+            logger.debug("no find_spec for finder %r", finder)
+            return
```
Yes, but wouldn't `getattr_static` work universally?

apotterri replied:
Oops, got the sense of your question, backwards. Sorry about that.

Yeah, `getattr_static` would probably work universally. It's just such a gross thing to have to do.... ;)

I'll check and make sure it works, but assuming it does, I'll just use that.

apotterri replied:
In fact, it [doesn't work](https://app.travis-ci.com/github/getappmap/appmap-python/jobs/602678808), so I'll just leave this as-is.

apotterri commented:
@dividedmind this is ready for another look

I made one fairly substantial change since you last looked: I removed the `wrapt` submodule, and replaced it by the use of the `vendoring` package. This seems to be a more [widely-used](https://github.com/pypa/pip/blob/main/pyproject.toml#L29) [approach](https://github.com/python-poetry/poetry-core/blob/5da9a2ed898d9438a8235169f3a4833a1380acdc/pyproject.toml#L117) to [vendoring](https://github.com/search?q=path%3A**%2Fpyproject.toml+tool.vendoring&type=code). One of its benefits is that it will allow us to easily vendor more packages, which we may need sooner rather than later: https://appmap-group.slack.com/archives/C03GM7D2QRH/p1684786853793949?thread_ts=1684779111.885699&cid=C03GM7D2QRH

dividedmind commented on README.md:
```diff
@@ -83,9 +92,9 @@ Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
-this easier to achieve, convention and refactoring checks have both been disabled. They
-should be reenabled as soon as possible.]
+[Note that the current configuration has a threshold set which must be met for the Travis build to
+pass. To make this easier to achieve, a number of checks have both been disabled. They should be
```
Good catch :)

dividedmind reviewed (approved):
Looks good, thanks for the changes! (I think I liked submoduling more, but if vendoring is easier to work with, that's fine.)
